iroh-gateway/src/core.rs

```diff
@@ -385,40 +366,143 @@ async fn serve_car(
         .client
         .get_file(
-            &req.full_content_path,
+            req.resolved_path.clone(),
             &state.rpc_client,
             start_time,
             Arc::clone(&state),
         )
         .await
         .map_err(|e| error(StatusCode::INTERNAL_SERVER_ERROR, &e, &state))?;
 
-    let name = add_content_disposition_headers(
-        &mut headers,
-        &req.query_file_name,
-        &req.content_path,
-        req.download,
-    );
-    set_etag_headers(&mut headers, get_etag(&req.cid, Some(req.format.clone())));
-    add_cache_control_headers(&mut headers, req.full_content_path.to_string());
-    add_content_type_headers(&mut headers, &name);
+    add_ipfs_roots_headers(&mut headers, metadata.clone());
+    match metadata.unixfs_type {
+        Some(UnixfsType::Dir) => {
+            if let Some(dir_list_data) = body.data().await {
+                let dir_list = match dir_list_data {
+                    Ok(b) => b,
+                    Err(_) => {
+                        return Err(error(
+                            StatusCode::INTERNAL_SERVER_ERROR,
+                            "failed to read dir listing",
+                            &state,
+                        ));
+                    }
+                };
+                return serve_fs_dir(&dir_list, req, state, headers, start_time).await;
+            } else {
+                return Err(error(
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    "failed to read dir listing",
+                    &state,
+                ));
+            }
+        }
+        Some(_) => {
+            // todo(arqu): error on no size
+            // todo(arqu): add lazy seeking
+            add_cache_control_headers(&mut headers, metadata.clone());
+            set_etag_headers(&mut headers, get_etag(&req.cid, Some(req.format.clone())));
+            let name = add_content_disposition_headers(
+                &mut headers,
+                &req.query_file_name,
+                &req.content_path,
+                req.download,
+            );
+            if metadata.unixfs_type == Some(UnixfsType::Symlink) {
+                headers.insert(
+                    CONTENT_TYPE,
+                    HeaderValue::from_str("inode/symlink").unwrap(),
+                );
+            } else {
+                add_content_type_headers(&mut headers, &name);
+            }
+        }
+        None => {
+            return Err(error(
+                StatusCode::BAD_REQUEST,
+                "couldn't determine unixfs type",
+                &state,
+            ));
+        }
+    }
     response(StatusCode::OK, body, headers)
 }
 
+#[tracing::instrument()]
+async fn serve_fs_dir(
+    dir_list: &Bytes,
+    req: &Request,
+    state: Arc<State>,
+    mut headers: HeaderMap,
+    start_time: std::time::Instant,
+) -> Result<GatewayResponse, GatewayError> {
+    let dir_list = String::from_utf8(dir_list.to_vec()).unwrap();
```

Comment:
you shouldn't need these allocs `std::str::from_utf8(&dir_list[..])`  should do the trick